### ISSUE TITLE
policy_eval function in Policy Iteration Solution.ipynb should use previous value funtion

### ISSUE THREAD

**Issue by DanTulovsky** [dennybritz/reinforcement-learning#101]
Hi

In reinforcement-learning/DP/Policy Iteration Solution.ipynb, in the policy_eval function, the value function used should be the value function of the previous state, not all zeros.  Please see page 97 in the text book.

Thank you
Dan

**Comment by ronaldseoh** [comment]
There is no previous value function when we start the policy evaluation.

**Comment by DanTulovsky** [comment]
I may be misunderstanding this.... but I think:

def policy_eval(policy, V, env, discount_factor=1.0, theta=0.00001):
  while True:
    ...

def policy_improvement(...):
  policy = ...
  V = np.zeros(env.nS)
  
  while True:
    V = policy_eval_fn(policy, V, env, discount_factor)

So you create the value functions once, at the beginning, not once for every policy_eval.  The book says:

"Note that each policy evaluation, itself an iterative computation, is started with the value function for the previous policy."

Dan

**Comment by ronaldseoh** [comment]
I was referring to V when I said "value function", because this is a special case of state value function where the function can be defined using a table consisting of values for every single state.  So we are not re-declaring the *Python* function each time we call policy_eval(). 
I haven't read the book yet, but you can see on page 83 that you will have to start with an array filled with zeros if you are only doing policy evaluation. However, for policy iteration, I can see that starting with the values from the previous policy can make the evaluation faster, since the new values would not have changed much from the one came right before. The policy evaluation algorithm will find the values for the current policy regardless of whether we start with previous policy values or not. Thank you for pointing this out!

**Comment by DanTulovsky** [comment]
Yes, sorry, I didn't mean to imply that the code was broken as written.  It
could just be made more efficient. :)

**Comment by Jaijith** [comment]
in the policy evaluation and policy iteration solution.ipynb; why is the value fuction calculated with the below equation.
                    v += action_prob * prob * (reward + discount_factor * V[next_state])

Shouldn't the value function be calculated as per the below equation
                     v += action_prob  * (reward +  prob*discount_factor * V[next_state])
The agent gets the reward as soon as it takes the action and the transistion probability is multiplied nly to the value function of the next state. 

Correct me if I am wronf

**Comment by honglonggiao** [comment]
It is totally right, Dan. We just need to initialize zero value function at the beginning. After that, we use the previous value functions to reduce the converging time.

**Comment by raslrn** [comment]
Hey guys, in policy iteration solution the policy_improvement function uses policy_eval as an input parameter. Running the code,  when policy_eval_fn is called the returned value (V) it's a matrix that reached the convergence rule delta < theta, consequently the policy_improvement perform policy improvement in this matrix. Is this correct? Should the improvement not be made after each evaluation?
